relentless/_collections.py

```diff
@@ -266,3 +268,171 @@ def __str__(self):
     def pairs(self):
         """tuple: All unique pairs in the matrix."""
         return tuple(self._data.keys())
+
+class KeyedArray(FixedKeyDict):
+    """Numerical array with fixed keys.
+
+    Can be used to perform arithmetic operations element-wise (per key) and
+    vector algebraic operations (norm, dot product).
+
+    Parameters
+    ----------
+    keys : array_like
+        List of keys to be fixed.
+    default : scalar
+        Initial value to fill in the dictionary, defaults to `None`.
+
+    Examples
+    --------
+    Create a keyed array::
+
+        k1 = KeyedArray(keys=('A','B'))
+        k2 = KeyedArray(keys=('A','B'))
+
+    Set values through update::
+
+        k1.update({'A':2.0, 'B':3.0})
+        k2.update({'A':3.0, 'B':4.0})
+
+    Perform vector arithmetic operations::
+
+        >>> print(k1 + k2)
+        {'A':5.0, 'B':7.0}
+        >>> print(k1 - k2)
+        {'A':-1.0, 'B':-1.0}
+        >>> print(3*k1)
+        {'A':6.0, 'B':9.0}
+        >>> print(k1/10)
+        {'A':0.2, 'B':0.3}
+
+    Compute vector dot product::
+
+        >>> print(k1.dot(k2))
+        18.0
+
+    Compute vector norm::
+
+        >>> print(k2.norm())
+        5.0
+
+    """
+    def __init__(self, keys, default=None):
+        super().__init__(keys, default)
+
+    def _assert_same_keys(self, val):
+        if (self.keys != val.keys):
+            raise KeyError('''Both KeyedArrays must have identical keys
+                              to perform mathematical operations.''')
+
+    def _assert_scalar(self, val):
+        if not np.isscalar(val):
+            raise TypeError('''A scalar is required for multiplication or
+                               division with a KeyedArray.''')
+
+    def __add__(self, val):
+        """Element-wise addition of two arrays."""
+        self._assert_same_keys(val)
+        k = KeyedArray(keys=self.keys)
+        k.update({x: self[x] + val[x] for x in self})
```

Comment:
For all of these methods, you might check if `val` is a `KeyedArray`, and if so, do element-wise addition. Otherwise, check if it is a scalar and just add that value. Otherwise, it is an error:
```py
if isinstance(val, KeyedArray):
    # element-wise
elif np.isscalar(val):
    # work as scalar
else:
    raise TypeError()
```